scripts/controllers/RobotArm.py

```diff
@@ -0,0 +1,198 @@
+        """
+
+        self.num_joints = num_joints
+        if controller_names is None:
+            controller_names = []
+            default_pos_names = ['panda_joint{}_position_controller'.format(i) for i in range(1, num_joints+1)]
+            default_vel_names = ['panda_joint{}_velocity_controller'.format(i) for i in range(1, num_joints+1)]
+            default_traj_name = ['panda_joint_trajectory_controller' if is_sim else 'joint_trajectory_controller']
+            controller_names.append(default_pos_names)
+            controller_names.append(default_vel_names)
+            controller_names.append(default_traj_name)
+
+        # create the controller manager for switching controllers.
+        self.controller_manager = RobotControllerManager(controller_names)
+
+        self.ignore_joints_arr = ignore_joints_arr
+        self.data_exists = False
+        self.joint_data = None
+        self.names = None
+        self.positions = None
+
+        # get list of position and velocity publishers ... and trajectory publisher.
+        self.pos_pubs, self.vel_pubs, self.traj_pub = self.get_publishers(controller_names)
+
+        rospy.Subscriber(joint_states_topic, JointState, self.joint_state_callback)
+        # need a sleep, so message is sent from publisher, interestingly.
+        rospy.sleep(2)
+
+    def joint_names(self):
+        """
+        get the joint names of the robot
+        """
+        if self.data_exists:
+            return self.names
+
+    def joint_angles(self):
+        """
+        gets the joint angles of the robot arm
+        """
+        if self.data_exists:
+            return np.array(self.joint_data.position)[self.valid_indices]
+
+    def joint_velocities(self):
+        """
+        gets the joint velocities of the robot arm
+        """
+        if self.data_exists:
+            return np.array(self.joint_data.velocity)[self.valid_indices]
+
+    def joint_angle(self, joint_name):
+        """
+        gets the joint angle based on the provided joint name.
+        """
+        if self.data_exists:
+            angles = self.joint_angles()
+            return angles[self.mapping[joint_name]]
+
+    def joint_velocity(self, joint_name):
+        """
+        gets the joint velocity based on the provided joint name.
+        """
+        if self.data_exists:
+            velocities = self.joint_velocities()
+            return velocities[self.mapping[joint_name]]
+
+    def set_joint_position_speed(self, speed=1.0):
+        rospy.logerr('Set Joint Position Speed Not Implemented for Robot Arm')
+
+    def move_to_joint_positions(self, positions, sleep=1.0, timeout=15.0):
+        """
+        moves robot to specific joint positions.
+        """
+        if not rospy.is_shutdown():
+            self.controller_manager.switch_mode(ControllerType.POSITION)
+            if len(positions) != self.num_joints:
+                raise ValueError("Wrong number of joints provided.")
```

Comment:
I would bring this at the top rather than including these lines in `if not rospy.is_shutdown():` block

Reply:
ah, makes sense

Reply:
updated